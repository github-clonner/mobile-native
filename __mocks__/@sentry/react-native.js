export const captureMessage = jest.fn();
export default {};
