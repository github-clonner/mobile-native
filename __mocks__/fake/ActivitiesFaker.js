
function load(count, container) {

  let activities = [...Array(count)].map((_, i) => {
    const code = 'activityguid' + i;
    return {
      attachment_guid: false,
      blurb: false,
      container_guid:code,
      custom_data:false,
      custom_type:false,
      rowKey: 'something' + i,
      description:"Congratulations! ",
      edited:"",
      guid:code,
      mature:false,
      time_created: "1522036284",
      ownerObj:{
        guid: "824853017709780997",
        type: "user",
        subtype: false,
        time_created: "1522036284",
        // getAvatarSource: () => {
        //   return {
        //     source:'http://thisisaurl'
        //   }
        // }
      },
      shouldBeBlured: jest.fn(),
      message:"Message",
      title:'TITLE',
      owner_guid:"824853017709780997",
      parent_guid:"838106762591510528",
      perma_url:false,
      thumbnail_src:false,
      type:"activity",
      wire_totals: {
        tokens: 1000000000000000000
      },
      containerObj: container
      // _list: {
      //   viewed: {
      //     viewed: new Map([["1019155171608096768",true]]),
      //     addViewed: () => {
      //       return;
      //     }
      //   }
      // },
      // getThumbSource: () => {
      //   return {
      //     source:'http://thisisaurl'
      //   }
      // }
    }
  });

  return {
    'activities': activities,
    'load-next': 'aaaaaa',
    'load-previous': 'aaaaaa'
  }

}

/**
 * comment service data faker
 * @param {integer} count
 */
export function activitiesServiceFaker() {
  return { load: load}
}

