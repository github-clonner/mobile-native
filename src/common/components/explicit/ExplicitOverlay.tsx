import React, { Component } from 'react';

import {
  Text,
  TouchableOpacity,
  StyleSheet,
  View,
  TextStyle,
  ViewStyle,
} from 'react-native';

import { observer } from 'mobx-react';
import { CommonStyle as CS } from '../../../styles/Common';
import Icon from 'react-native-vector-icons/MaterialIcons';
import { GOOGLE_PLAY_STORE } from '../../../config/Config';
import i18n from '../../services/i18n.service';
import type ActivityModel from 'src/newsfeed/ActivityModel';

type PropsType = {
  entity: ActivityModel;
  iconSize: number;
  hideText: boolean;
  iconPosition: 'right' | 'left';
  fontStyle?: TextStyle | Array<TextStyle>;
  containerStyle?: ViewStyle | Array<ViewStyle>;
  closeContainerStyle?: ViewStyle | Array<ViewStyle>;
};

/**
 * Explicit overlay
 */
@observer
export default class ExplicitOverlay extends Component<PropsType> {
  /**
   * Default props
   */
  static defaultProps = {
    iconSize: 80,
    hideText: false,
    iconPosition: 'right',
  };

  /**
   * toggle overlay
   */
  toogle = () => {
    this.props.entity.toggleMatureVisibility();
  };

  /**
   * Render
   */
  render() {
    const {
      iconSize,
      hideText,
      fontStyle,
      containerStyle,
      closeContainerStyle,
      iconPosition,
    } = this.props;

    if (this.props.entity.mature_visibility) {
      const size = Math.max(iconSize / 2.5, 25);
      const styles =
        iconPosition === 'right'
          ? [CS.positionAbsoluteTopRight, CS.paddingRight3x, CS.marginRight3x]
          : [CS.positionAbsoluteTop, CS.paddingLeft3x, CS.marginLeft3x];
      return (
        <View style={[styles, closeContainerStyle]}>
          <Icon
            name="explicit"
            size={size}
            color={'red'}
            style={CS.shadow}
            onPress={this.toogle}
          />
        </View>
      );
    }

    const text = GOOGLE_PLAY_STORE
      ? i18n.t('postCantBeShown')
      : i18n.t('confirm18');

    return (
      <TouchableOpacity
        activeOpacity={1}
        style={[
          CS.positionAbsolute,
          CS.centered,
          CS.backgroundDarkGreyed,
          styles.onTop,
          containerStyle,
        ]}
        onPress={this.toogle}>
        <Icon
          name="explicit"
          size={iconSize}
          color={'white'}
          style={CS.shadow}
        />
        {!hideText && (
          <Text style={[CS.colorWhite, CS.shadow, fontStyle]}>{text}</Text>
        )}
      </TouchableOpacity>
    );
  }
}

const styles = StyleSheet.create({
  onTop: {
    zIndex: 99999,
  },
});
