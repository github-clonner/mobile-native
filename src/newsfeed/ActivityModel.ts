import { runInAction, action, observable, decorate } from 'mobx';
import FastImage from 'react-native-fast-image';
import { FlatList, Alert } from 'react-native';

import BaseModel from '../common/BaseModel';
import UserModel from '../channel/UserModel';
import wireService from '../wire/WireService';
import sessionService from '../common/services/session.service';
import {
  setPinPost,
  deleteItem,
  unfollow,
  follow,
  update,
} from '../newsfeed/NewsfeedService';
import api from '../common/services/api.service';

import { GOOGLE_PLAY_STORE, MINDS_CDN_URI, MINDS_URI } from '../config/Config';
import i18n from '../common/services/i18n.service';
import logService from '../common/services/log.service';
import entitiesService from '../common/services/entities.service';
import type { ThumbSize } from '../types/Common';
import type GroupModel from 'src/groups/GroupModel';

type Thumbs = Record<ThumbSize, string>;

/**
 * Activity model
 */
export default class ActivityModel extends BaseModel {
  // Observable properties
  @observable pinned: boolean = false;
  @observable time_created: string = '';
  @observable message: string = '';
  @observable title: string = '';
  @observable mature: boolean = false;
  @observable edited: '0' | '1' = '0';

  // decorated observables
  'is:following': boolean;
  'thumbs:down:count': number;
  'thumbs:up:count': number;
  'comments:count': number;
  'thumbs:down:user_guids': Array<number>;
  'thumbs:up:user_guids': Array<number>;

  containerObj?: GroupModel;
  remind_object?: ActivityModel;
  ownerObj!: UserModel;
  listRef?: FlatList<any>;
  thumbnails?: Thumbs;
  paywall: '1' | '' = '';
  paywall_unlocked: boolean = false;
  guid: string = '';
  subtype: string = '';
  entity_guid: string = '';
  owner_guid: string = '';
  custom_type?: string;
  custom_data?: Array<any> | any;
  nsfw?: Array<number>;
  flags?: any;
  reminds: number = 0;
  impressions: number = 0;
  perma_url?: string;
  cinemr_guid?: string;
  thumbnail_src?: string;
  dontPin?: boolean;
  boosted?: boolean;
  wire_threshold?: {
    type: 'token' | 'money';
    min: number;
  } | null;
  _preview?: boolean;
  /**
   * Set programatically to indicate that the parent is marked as mature
   */
  is_parent_mature?: boolean;

  /**
   * Mature visibility flag
   */
  @observable mature_visibility: boolean = false;

  /**
   * Is visible in flat list
   */
  @observable is_visible: boolean = true;

  /**
   *  List reference setter
   */
  set _list(value) {
    this.__list = value;

    // the reminded object need access to the metadata service too
    if (this.remind_object) {
      this.remind_object._list = value;
    }
  }

  /**
   *  List reference getter
   */
  get _list() {
    return this.__list;
  }

  toPlainObject() {
    const plainEntity = super.toPlainObject();
    if (plainEntity.remind_object && plainEntity.remind_object.__list) {
      delete plainEntity.remind_object.__list;
    }

    return plainEntity;
  }

  /**
   * Child models
   */
  childModels() {
    return {
      ownerObj: UserModel,
      remind_object: ActivityModel,
    };
  }

  /**
   * Get the activity thumb source
   * {uri: 'http...'}
   * @param {string} size
   */
  getThumbSource(size: ThumbSize = 'medium') {
    // for gif use always the same size to take adventage of the cache (they are not resized)
    if (this.isGif()) {
      size = 'medium';
    }

    if (this.thumbnails && this.thumbnails[size]) {
      return { uri: this.thumbnails[size], headers: api.buildHeaders() };
    }

    // fallback to old behavior
    if (this.paywall || this.paywall_unlocked) {
      return {
        uri: MINDS_URI + 'fs/v1/thumbnail/' + this.entity_guid,
        headers: api.buildHeaders(),
      };
    }
    if (this.custom_type === 'batch') {
      return {
        uri: MINDS_CDN_URI + 'fs/v1/thumbnail/' + this.entity_guid + '/' + size,
        headers: api.buildHeaders(),
      };
    }
    return {
      uri: MINDS_CDN_URI + 'fs/v1/thumbnail/' + this.guid + '/' + size,
      headers: api.buildHeaders(),
    };
  }

  isGif(): boolean {
    return this.custom_data && this.custom_data[0] && this.custom_data[0].gif;
  }

  /**
   * Preload thumb on image cache
   */
  preloadThumb(size: ThumbSize = 'medium') {
    FastImage.preload([this.getThumbSource(size)]);
  }

  shouldBeBlured(): boolean {
    const user = sessionService.getUser();

    if (user && user.mature) {
      return false;
    }

    if (this.nsfw !== undefined) {
      let res = [1, 2, 4].filter((nsfw) => this.nsfw!.indexOf(nsfw) > -1)
        .length;
      if (res) {
        return true;
      }
    }

    if (typeof this.flags !== 'undefined') {
      return !!this.flags.mature;
    }

    if (typeof this.mature !== 'undefined') {
      return !!this.mature;
    }

    if (
      typeof this.custom_data !== 'undefined' &&
      typeof this.custom_data[0] !== 'undefined'
    ) {
      return !!this.custom_data[0].mature;
    }

    if (this.custom_data !== undefined) {
      return !!this.custom_data.mature;
    }

    return false;
  }

  /**
   * Get activity text
   */
  get text() {
    return this.message || '';
  }

  @action
  toggleMatureVisibility() {
    if (GOOGLE_PLAY_STORE) {
      return;
    }
    this.mature_visibility = !this.mature_visibility;

    if (this.remind_object && this.remind_object.mature) {
      this.remind_object.mature_visibility = this.mature_visibility;
    }
  }

  @action
  setVisible(value: boolean) {
    this.is_visible = value;
  }

  /**
   * Increment the comments counter
   */
  @action
  incrementCommentsCounter() {
    this['comments:count']++;
  }

  /**
   * Decrement the comments counter
   */
  @action
  decrementCommentsCounter() {
    this['comments:count']--;
  }

  /**
   * Unlock the activity and update data on success
   */
  @action
  async unlock(ignoreError = false) {
    try {
      const result: object | false = await wireService.unlock(this.guid);
      if (result) {
        // all changes should be atomic (trigger render only once)
        runInAction(() => {
          // create a new model because we need the child models
          const model: ActivityModel = ActivityModel.create(result);
          Object.assign(this, model);
        });
      }
      return result;
    } catch (err) {
      if (!ignoreError) {
        Alert.alert(err.message);
      }
      return false;
    }
  }

  @action
  async togglePin() {
    // allow owners only
    if (!this.isOwner()) {
      return;
    }

    try {
      this.pinned = !this.pinned;
      await setPinPost(this.guid, this.pinned);
    } catch (e) {
      this.pinned = !this.pinned;
      Alert.alert(i18n.t('errorPinnedPost'));
    }
  }

  @action
  async deleteEntity() {
    try {
      await deleteItem(this.guid);
      this.removeFromList();
      entitiesService.deleteFromCache(this.urn);
    } catch (err) {
      logService.exception('[ActivityModel]', err);
      throw err;
    }
  }

  @action
  async toggleFollow() {
    const method = this['is:following'] ? unfollow : follow;
    try {
      await method(this.guid);
      runInAction(() => {
        this['is:following'] = !this['is:following'];
      });
    } catch (err) {
      logService.exception('[OffsetFeedListStore]', err);
      throw err;
    }
  }

  @action
  async updateActivity(data: any = {}) {
    const entity: any = this.toPlainObject();

    if (data) {
      for (const field in data) {
        entity[field] = data[field];
      }
    }

    // call update endpoint
    await update(entity);

    // update instance properties
    this.update(data);

    this.setEdited(entity.message);
  }

  @action
  setEdited(message: string) {
    this.message = message;
    this.edited = '1';
  }
}

/**
 * Define model observables
 */
decorate(ActivityModel, {
  //@ts-ignore
  'thumbs:down:count': observable,
  'thumbs:up:count': observable,
  'comments:count': observable,
  'thumbs:down:user_guids': observable,
  'thumbs:up:user_guids': observable,
});
