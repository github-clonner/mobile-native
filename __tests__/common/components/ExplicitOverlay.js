import 'react-native';
import React from 'react';
import { shallow } from 'enzyme';
import ExplicitOverlay from '../../../src/common/components/explicit/ExplicitOverlay';
import ActivityModel from '../../../src/newsfeed/ActivityModel';
import { activitiesServiceFaker } from '../../../__mocks__/fake/ActivitiesFaker';

describe('Explicit overlay component', () => {

  let entity, screen;
  beforeEach(() => {

    let mockResponse = activitiesServiceFaker().load(1);
    entity = ActivityModel.create(mockResponse.activities[0]);
    entity.mature_visibility = false;
    screen = shallow(
      <ExplicitOverlay entity={entity}/>
    );
  });

  it('renders correctly', async () => {
    expect(screen).toMatchSnapshot();
  });

  it('should renders the etag', async () => {
    entity.mature_visibility = true;
    await screen.update();

    expect(screen).toMatchSnapshot();
  });
});
