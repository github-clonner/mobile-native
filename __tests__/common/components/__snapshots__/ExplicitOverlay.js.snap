// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Explicit overlay component renders correctly 1`] = `
<TouchableOpacity
  activeOpacity={1}
  onPress={[Function]}
  style={
    Array [
      Object {
        "bottom": 0,
        "left": 0,
        "position": "absolute",
        "right": 0,
        "top": 0,
      },
      Object {
        "alignContent": "center",
        "alignItems": "center",
        "alignSelf": "center",
        "justifyContent": "center",
      },
      Object {
        "backgroundColor": "#777777",
      },
      Object {
        "zIndex": 99999,
      },
      undefined,
    ]
  }
>
  <Icon
    allowFontScaling={false}
    color="white"
    name="explicit"
    size={80}
    style={
      Object {
        "elevation": 4,
        "shadowColor": "black",
        "shadowOffset": Object {
          "height": 2,
          "width": 0,
        },
        "shadowOpacity": 0.1,
        "shadowRadius": 2,
      }
    }
  />
  <Text
    style={
      Array [
        Object {
          "color": "#FFFFFF",
        },
        Object {
          "elevation": 4,
          "shadowColor": "black",
          "shadowOffset": Object {
            "height": 2,
            "width": 0,
          },
          "shadowOpacity": 0.1,
          "shadowRadius": 2,
        },
        undefined,
      ]
    }
  >
    Confirm you are 18+
  </Text>
</TouchableOpacity>
`;

exports[`Explicit overlay component should renders the etag 1`] = `
<View
  style={
    Array [
      Array [
        Object {
          "position": "absolute",
          "right": 0,
          "top": 0,
        },
        Object {
          "paddingRight": 15,
        },
        Object {
          "marginRight": 15,
        },
      ],
      undefined,
    ]
  }
>
  <Icon
    allowFontScaling={false}
    color="red"
    name="explicit"
    onPress={[Function]}
    size={32}
    style={
      Object {
        "elevation": 4,
        "shadowColor": "black",
        "shadowOffset": Object {
          "height": 2,
          "width": 0,
        },
        "shadowOpacity": 0.1,
        "shadowRadius": 2,
      }
    }
  />
</View>
`;
